add flash get size API function for the stm32 devices

FRASTM opened the pull request:
Add the support of the API function **flash_get_size** for the stm32 mcu devices

According to the [documentation](https://docs.zephyrproject.org/latest/doxygen/html/group__flash__interface.html#ga471edb53a2b1abe3dff0f4e2d216521e
), the size is in expressed in Bytes.
The stm32Cube LL function LL_GetFlashSize  is directly giving the size in Kbytes from the mcu register "FLASHSIZE_BASE" 

The tests/drivers/flash/common/ is modified to accept a testcase for all the stm32 targets when the 
CONFIG_TEST_DRIVER_FLASH_SIZE <> 1

```
TESTSUITE flash_driver succeeded                                                                        
                                                                                                        
------ TESTSUITE SUMMARY START ------                                                                   
                                                                                                        
SUITE PASS - 100.00% [flash_driver]: pass = 7, fail = 0, skip = 0, total = 7 duration = 0.370 seconds
 - PASS - [flash_driver.test_flash_copy] duration = 0.119 seconds
 - PASS - [flash_driver.test_flash_erase] duration = 0.055 seconds
 - PASS - [flash_driver.test_flash_fill] duration = 0.051 seconds
 - PASS - [flash_driver.test_flash_flatten] duration = 0.073 seconds
 - PASS - [flash_driver.test_flash_page_layout] duration = 0.032 seconds
 - PASS - [flash_driver.test_get_size] duration = 0.005 seconds
 - PASS - [flash_driver.test_read_unaligned_address] duration = 0.035 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```


requires https://github.com/zephyrproject-rtos/zephyr/pull/83144

FRASTM committed:
drivers: flash: stm32 flash drivers supports the get_size API function

Implement the get_size flash device API function for the various stm32
This is just CONFIG_FLASH_SIZE expressed in Bytes or the FLASH_SIZE
given by the LL driver

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
CI failure  when building samples/bluetooth/peripheral/ with nucleo_wb55rg : see issue https://github.com/zephyrproject-rtos/zephyr/issues/82344

FRASTM commented:
waiting for the PR https://github.com/zephyrproject-rtos/zephyr/pull/88585

FRASTM committed:
tests: drivers: flash testing stm32 flash size

Add a special testcase to test any stm32 target
The expected flash size is given by the config FLASH_SIZE (in bytes)
and stm32 flash driver API function should return the same value
Do not run on the stm32wb55 nucleo as 216K are reserved among the 1MB
(only 808K declared but mcu has 1MB)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Rrebase on c9f149cf6f445d7e39681ed0509e7acab5dee25b
In case of a **stm32 device**, which now has the get_size API, the testcase is adapting the test case and sets the CONFIG_TEST_DRIVER_FLASH_SIZE to the value given by the reg size of the flash0 node.  Instead of adding one _.conf_ file per device like CONFIG_TEST_DRIVER_FLASH_SIZE=2097152
That could be extended to any other soc vendor.

de-nordic commented on tests/drivers/flash/common/Kconfig:
```diff
@@ -4,11 +4,15 @@
 	int "Size of flash device under test"
-	default -1
+	default -1 if !SOC_FAMILY_STM32
+	default $(dt_chosen_reg_size_int,$(DT_CHOSEN_Z_FLASH),0) if SOC_FAMILY_STM32
```
Just move the line above `default -1`, Kconfig will pick first default that has condition evaluated to true.

etienne-lms commented on drivers/flash/flash_stm32.c:
```diff
@@ -431,6 +432,16 @@ flash_stm32_get_parameters(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	*size = (uint64_t)(LL_GetFlashSize() * 1024);
```
I guess we're not dealing with Terabyte flashes yet. That said, i could be better to cast `LL_GetFlashSize()` return value before the multiplication. 
By the way, I think `1024` should be an explicit unsigned value. 
Suggestion:
```c
	*size = (uint64_t)LL_GetFlashSize() * 1024U;

FRASTM committed:
drivers: flash: stm32H5 flash driver gives the flash size

Use the LL_GetFlashSize to retrieve the size of the flash, reading the
system flash OTP of the stm32h5 serie. This operation requires
Icache disable/re-enable to access the 0x08FF F80C memory area

Signed-off-by: Francois Ramu <francois.ramu@st.com>

etienne-lms commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -861,6 +861,16 @@ static const struct flash_parameters *flash_stm32h7_get_parameters(const struct
+{
+	ARG_UNUSED(dev);
+
+	*size = (uint64_t)(LL_GetFlashSize() * 1024);
```
Sorry, I was not explicit. My comment applies also here and in **drivers/flash/flash_stm32wb0x.c** and **drivers/flash/flash_stm32wba_fm.c**.

FRASTM replied:
Ok, modified foe all other series

de-nordic commented on tests/drivers/flash/common/Kconfig:
```diff
@@ -4,11 +4,15 @@
 	int "Size of flash device under test"
-	default -1
+	default $(dt_chosen_reg_size_int,$(DT_CHOSEN_Z_FLASH),0) if SOC_FAMILY_STM32
+	default -1 if !SOC_FAMILY_STM32
```
Bring back the original here, the change is no longer needed.

FRASTM replied:
Ok, this is now:
```
	default -1 if !SOC_FAMILY_STM32
	default $(dt_chosen_reg_size_int,$(DT_CHOSEN_Z_FLASH),0) if SOC_FAMILY_STM32
```

de-nordic commented on tests/drivers/flash/common/Kconfig:
```diff
@@ -4,11 +4,15 @@
 config TEST_DRIVER_FLASH_SIZE
 	int "Size of flash device under test"
-	default -1
+	default -1 if !SOC_FAMILY_STM32
+	default $(dt_chosen_reg_size_int,$(DT_CHOSEN_Z_FLASH),0) if SOC_FAMILY_STM32
```
Oh, no! That what I have meant:
```suggestion	
	default $(dt_chosen_reg_size_int,$(DT_CHOSEN_Z_FLASH),0) if SOC_FAMILY_STM32
	default -1
```
Sorry for confusion.

FRASTM replied:
Ok, I see
Done

FRASTM commented on drivers/flash/flash_stm32.c:
```diff
@@ -431,6 +436,29 @@ flash_stm32_get_parameters(const struct device *dev)
+	 * This is required on STM32H5, where the manufacturing flash must be
+	 * accessed in non-cacheable mode - otherwise, a bus error occurs.
+	 */
+	LL_ICACHE_Disable();
```
done

FRASTM commented on drivers/flash/flash_stm32.c:
```diff
@@ -431,6 +436,29 @@ flash_stm32_get_parameters(const struct device *dev)
+
+#if defined(CONFIG_SOC_SERIES_STM32H5X)
+	/* Re-enable the ICACHE (unconditonally - it should always be turned on) */
+	LL_ICACHE_Enable();
```
done
